File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        mult = 1
        res = []

        zeros = []
        for i, num in enumerate(nums):
            if num == 0:
                zeros.append(i) 
                continue
            mult*=num
            
        if len(zeros)>=2:
            return [0]*len(nums)
        elif len(zeros)>=1:
            return [0]*(zeros[0])+ [mult] + [0]*(len(nums)-zeros[0]-1)
 
        for i, num in enumerate(nums, 1):
            if num == 0: 
                res.append(mult)
                continue                
            n = int(mult*(num**-1))
            res.append(n)
            
        return res

File: 238-product-of-array-except-self/product-of-array-except-self.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let mut prefix: Vec<i32>= Vec::<i32>::new();
        let mut suffix: Vec<i32> = Vec::<i32>::new();
        let mut res: Vec<i32> = Vec::<i32>::new();
        let mut p:i32= 1;
        let mut s:i32 = 1;
        let n: i32 = nums.len() as i32;
        for i in 0..n {
            prefix.push(p);
            suffix.push(s);
            
            p = p*nums[i as usize];
            s = s*nums[(n-i-1) as usize];
        }
        for i in 0..n{
            res.push(prefix[i as usize]*suffix[(n-i-1) as usize]);
        }

        return res;
    }
}